### DISCUSSION TITLE
Questions on Versioning and Dependencies for .NET Components: SDK vs. NuGet Package Versions

### DISCUSSION THREAD

**Discussion by noSet** [dotnet/runtime#110265] [Q&A] [answered]
##### Question 1
I'm developing a component. This component depends on the `Microsoft.Extensions.DependencyInjection.Abstractions` package. My target framework might be `.NET Framework 4.8`, `.NET 6.0`, and `.NET 8.0`. Should I reference different versions of `Microsoft.Extensions.DependencyInjection.Abstractions` for each target framework?

For example:

```xml
<PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.0.0" Condition="'$(TargetFramework)' == 'net6.0' "/>
<PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="8.0.0" Condition="'$(TargetFramework)' == 'net8.0' "/>
<PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="?.0.0" Condition="'$(TargetFramework)' == 'net48' "/>
```

##### Question 2
If I'm developing a component targeting **only** the .NET 6.0 framework, should I reference `Microsoft.Extensions.DependencyInjection.Abstractions` version `6.0.0` or `8.0.0`? I can find version `8.0.0` in NuGet Package Manager.

**Choosing SDK versions and NuGet package versions is really confusing and troublesome for me.**

**Comment by huoyaoyuan** [comment, marked as answer]
> Should I reference different versions of `Microsoft.Extensions.DependencyInjection.Abstractions` for each target framework?

No, this is wrong. You should reference the *same* version of package, usually the latest one. Out-of-band NuGet packages are meant to be used with different versions of framework. The version number means the timeline of the package produced, but they are not tied to the framework version.
Rarely, these packages may contain breaking change between versions. Using the same package version helps you ensuring the same feature and behavior.

> If I'm developing a component targeting **only** the .NET 6.0 framework, should I reference `Microsoft.Extensions.DependencyInjection.Abstractions` version `6.0.0` or `8.0.0`?

Use the latest version that provides support on the framework, in this case you should use `8.0.2`, since `9.0.0` is dropping support for .NET 6. Using latest patch is also important.

> Choosing SDK versions and NuGet package versions is really confusing and troublesome for me.

You are probably messed up with package version and [supported framework version](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection.Abstractions#supportedframeworks-body-tab). The package version number itself isn't functionally meaningful.
